spring-integration-event/src/main/java/org/springframework/integration/event/inbound/ApplicationEventListeningMessageProducer.java

```diff
@@ -116,20 +129,45 @@ public void onApplicationEvent(ApplicationEvent event) {
 		}
 	}
 
+	@Override
+	protected Object evaluatePayloadExpression(Object rootObject) {
+		if (!this.payloadExpressionSet && rootObject instanceof PayloadApplicationEvent) {
+			return ((PayloadApplicationEvent) rootObject).getPayload();
```

Comment:
Well, the Javadoc may be confusing. Let me know if we can improve that. In practice you should not worry at all about that stuff on the publishing side (you're just publishing an object and we'll make sure to wrap it if necessary). I am not entirely sure why you actually need to bother about it to be honest.